math/vecops/inc/ROOT/RVec.hxx

```diff
@@ -680,14 +680,61 @@ RVec<typename RVec<T>::size_type> Argsort(const RVec<T> &v)
 
 /// Return elements of a vector at given indices
 template <typename T>
-RVec<T> ByIndices(const RVec<T> &v, const RVec<typename RVec<T>::size_type> &i)
+RVec<T> TakeIndices(const RVec<T> &v, const RVec<typename RVec<T>::size_type> &i)
 {
    RVec<T> r(i.size());
    for (unsigned int k = 0; k < i.size(); k++)
       r[k] = v[i[k]];
    return r;
 }
 
+/// Return first elements of a vector
+template <typename T>
+RVec<T> TakeFirst(const RVec<T> &v, const typename RVec<T>::size_type n)
+{
+   RVec<T> r(n);
+   for (typename RVec<T>::size_type k = 0; k < n; k++)
+      r[k] = v[k];
+   return r;
+}
+
+/// Return last elements of a vector
+template <typename T>
+RVec<T> TakeLast(const RVec<T> &v, const typename RVec<T>::size_type n)
+{
+   RVec<T> r(n);
+   for (typename RVec<T>::size_type k = 0; k < n; k++)
+      r[k] = v[v.size() - n + k];
```

Comment:
you are calling `size()` at every iteration here, and things go badly if `n > size`. What about something like
```c++
RVec<T> r;                                                                                                 
r.reserve(n);                                                                                                       
const auto size = v.size();                                                                                         
for (typename RVec<T>::size_type k = std::max(size - n, 0u); k < size; ++k) // ++k avoids the creation of a temporary w.r.t. k++
   r.emplace_back(v[k]);
```

Reply:
i tried to benchmark the performance difference of the two versions.
i'm getting two completely different results with [gcc8](http://quick-bench.com/pmXdMY8fs6pVka6WZoXa9E_FD-M) and with [clang](http://quick-bench.com/Ai6qbXEQDD-U2g2HWASgIoK83Lo).

A version that does not use `reserve` but still avoids calling `size()` at every iteration [seems to work best](http://quick-bench.com/OTdqMrlMLsrvDuvJliVi2a_O-oc) overall